### ISSUE TITLE
[Question] What is the correct way to create a spec file that shares a browserContext

### ISSUE THREAD

**Issue by mkiddTempus** [microsoft/playwright#9034]
I'm confused as to how to share a browser context within a single spec file. I'm basing my usage of describe, beforeAll, beforeEach, and test on previous frameworks so completely understand if I'm setting this up wrong. 

My ultimate goal is within a SINGLE spec file, I would have a test set up like the following to allow for test separation and execution via having applied a localStorage to the shared browser context to address authentication. (apologies for formatting of pseudo code, PREVIOUSLY is what I was doing):

```
test.describe("Test description", () => {
	let context;
	let authToken;

	test.beforeAll(async ({ browser }) => {
		// PREVIOUSLY: retrieve token for test set up
		authToken = retrieveToken();
		// PREVIOUSLY: use this token to make API calls for data set up

		// IDEAL GOAL: Establish single browser context and apply the created storageState
		// browser = await chromium.launch();
		// context = await browser.newContext({ storageState: localStorageBuilt });
	}
	test.beforeEach(async ({ page }, testInfo) => {
		// PREVIOUSLY: opened a browser to the authentication system to retrieve the token implicitly in the created page
		// IDEAL GOAL: uncertain given the intended set up, may not need to do anything.
	}
	test("Test 1", async ({ page }, testInfo) => {
		// PREVIOUSLY: reused the page (and I'm assuming the implicitly created browserContext associated to the page), so the token was in the localStorage
		// IDEAL GOAL: Create new page to start test from already established single browser context to allow for all 3 to tests to run at once.
	}
	test("Test 2", async ({ page }, testInfo) => {
		// PREVIOUSLY: commented out so does not run while iterating on development
		// Ideal goal: Create new page to start test from already established single browser context to allow for all 3 to tests to run at once.
	}
	test("Test 3", async ({ page }, testInfo) => {
		// PREVIOUSLY: commented out so does not run while iterating on development
		// Ideal goal: Create new page to start test from already established single browser context to allow for all 3 to tests to run at once.
	}
}

```

**Comment by mxschmitt** [comment]
If you want to share a single browser context, then this would be something for you: https://playwright.dev/docs/test-parallel#reusing-single-page-between-tests

By default Playwright gives you a fresh browser context for each to ensure test isolation etc.

**Comment by klhex** [comment]
@mxschmitt In PW v1.15 that part of the documentation has moved to here: https://playwright.dev/docs/test-retries#reuse-single-page-between-tests 😉

**Comment by mkiddTempus** [comment]
Question: Is `test.describe.parallel` only available in v1.14 or is it only available in v1.15?`

This is starting to make more sense as to what you have available to you with the test framework, I think.

In beforeAll you have access to the browser. Makes sense since you would tracing across all the files. I would think this is where you would set up the localStorage so that all pages in the tests have the localStorage set when the begin execution. 


However you can get to WHERE via `page.context() -> browserContext.browser() -> browser.newContext()` but I don't see a way that this context gets reused or how the defaultContext is created if I were to want the tests to run in parallel (or serially since I could set each tests data all in the beforeAll.) 

Am I supposed to reassign a new context to the page in the beforeEach or in each test?

**Comment by mxschmitt** [comment]
> Question: Is test.describe.parallel only available in v1.14 or is it only available in v1.15?`

it got introduced in version 1.15 which got released yesterday.

> I would think this is where you would set up the localStorag

Usually you have by default a new contxt per each test. Which is isolated, so you would have to use beforeEach instead. If you are using describe.serial() from what I shared above, then you can use beforeAll and create your own context there.

The default context is created automatically before each test and closed after each test.

Usually you don't need test.describe.parallel. Playwright runs the tests per file in parallel by default.

> Am I supposed to reassign a new context to the page in the beforeEach or in each test?

**Comment by mxschmitt** [comment]
> Question: Is test.describe.parallel only available in v1.14 or is it only available in v1.15?`

it got introduced in version 1.15 which got released yesterday.

> I would think this is where you would set up the localStorag

Usually you have by default a new contxt per each test. Which is isolated, so you would have to use beforeEach instead. If you are using describe.serial() from what I shared above, then you can use beforeAll and create your own context there.

The default context is created automatically before each test and closed after each test.

Usually you don't need test.describe.parallel. Playwright runs the tests per file in parallel by default.

> Am I supposed to reassign a new context to the page in the beforeEach or in each test?

Depends.

a) you don't use .serial or .parallel -> context isolation -> no need to create context, or page manually. You need to use beforeEach if you want to change localStorage
b) you use .serial() -> you have access in beforeAll to the page, and context. -> it stops the following tests in the file after the first test failure.
c) you use test.describe -> its basically a) just in parallel.

We recommend a). But for folks with big test workflows b) might be useful.

Let me know if you have more concrete questions.

**Comment by mxschmitt** [comment]
Closing since the issue seems answered. For any other questions please open a new one.
